### PR TITLE
expose more diagnostic info in console when an actor throws an exception

### LINKED ISSUE #1187: If there is a jinja error you don't get any error message
I'm hacking on a track change now and getting messages like:
```
[ERROR] Cannot race. Error in race control (Could not load '/home/nik9000/Code/Elastic/rally-tracks/http_logs/track.json'. The complete track has been written to '/tmp/tmpo8b2g1ng.json' for diagnosis.)
```
There isn't a `/tmp/tmpo8b2g1ng.json`. I'm fairly sure this is because I have a syntax error in my jinja somewhere and jinja is bailing before it can write the file. I *think* jinja should have an error message telling me what I'm doing wrong. But I *think* we're eating the error message and not reporting it. Can we?

### PR DESCRIPTION
**Pull request by DJRickyB**
Resolves #1187 

With this commit, we expose more information to the console to help troubleshoot a failed command.  For example, if one changes `runtime_fields` to `runtime-fields` in https://github.com/elastic/rally-tracks/blob/master/http_logs/track.json#L48, a Jinja error occurs.  Previously, the console output would look like this:

```
 % esrally race --track-path ../rally-tracks/http_logs --test-mode

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [51e50d00-8d63-4ee6-9c5c-2f8757871a82]
[ERROR] Cannot race. Error in race control (Could not load '/Users/rick.boyd/dev/rally-tracks/http_logs/track.json'. The complete track has been written to '/var/folders/_j/86s8k_jd4md2k6ch_290_0mr0000gn/T/tmpippzkj_l.json' for diagnosis.)

Getting further help:
*********************
* Check the log files in /Users/rick.boyd/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/latest/.
* Ask a question on the forum at https://discuss.elastic.co/tags/c/elastic-stack/elasticsearch/rally.
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /Users/rick.boyd/.rally/logs.

-------------------------------
[INFO] FAILURE (took 4 seconds)
-------------------------------
```

With this change, the output is more descriptive:
```
 % esrally race --track-path ../rally-tracks/http_logs --test-mode     

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [df2c8410-83e9-4cd0-9c30-aa15d84a03bf]
[ERROR] Cannot race. esrally.track.loader.TrackSyntaxError: Could not load '/Users/rick.boyd/dev/rally-tracks/http_logs/track.json'. The complete track has been written to '/var/folders/_j/86s8k_jd4md2k6ch_290_0mr0000gn/T/tmp48mzhxww.json' for diagnosis. (Caused by: jinja2.exceptions.UndefinedError: 'runtime' is undefined)

Getting further help:
*********************
* Check the log files in /Users/rick.boyd/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/latest/.
* Ask a question on the forum at https://discuss.elastic.co/tags/c/elastic-stack/elasticsearch/rally.
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /Users/rick.boyd/.rally/logs.

-------------------------------
[INFO] FAILURE (took 4 seconds)
-------------------------------


```

### CODE REVIEW

**Comment by pquentin** [comment]
Nice! That's my main frustration with the actor system. Can we also print a stack trace using [traceback.format_exc](https://docs.python.org/3/library/traceback.html#traceback.format_exc)?

**Comment by DJRickyB** [comment]
I'd rather the stack trace remained in the log, perhaps we can put a static mention in the error message like `Please see the logs for more details`?

**Comment by pquentin** [comment]
We can leave it in the logs if you prefer, but the one that's currently in the logs isn't useful as it's outside the actor system

**Comment by DJRickyB** [comment]
> We can leave it in the logs if you prefer, but the one that's currently in the logs isn't useful as it's outside the actor system

I'm not sure I understand.  Are you suggesting the stacktrace doesn't populate in rally.log due to the bounds of the actors? This does not seem to be the case. For instance:

<details><summary>rally.log</summary>

```
2021-09-22 11:45:23,468 ActorAddr-(T|:64996)/PID:77519 esrally.actor ERROR Error in race control
Traceback (most recent call last):

  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 982, in read
    rendered = render_template_from_file(track_spec_file, self.track_params, complete_track_params=self.complete_track_params)

  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 781, in render_template_from_file
    return render_template(

  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 750, in render_template
    return template.render()

  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()

  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))

  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)

  File "<template>", line 48, in top-level template code

jinja2.exceptions.UndefinedError: 'runtime' is undefined


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/Users/rick.boyd/dev/rally/esrally/actor.py", line 96, in guard
    return f(self, msg, sender)

  File "/Users/rick.boyd/dev/rally/esrally/racecontrol.py", line 110, in receiveMsg_Setup
    self.coordinator.setup(sources=msg.sources)

  File "/Users/rick.boyd/dev/rally/esrally/racecontrol.py", line 194, in setup
    self.current_track = track.load_track(self.cfg)

  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 192, in load_track
    return _load_single_track(cfg, repo, repo.track_name)

  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 199, in _load_single_track
    current_track = reader.read(track_name, track_repository.track_file(track_name), track_dir)

  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 1008, in read
    raise TrackSyntaxError(msg, str(e))

esrally.track.loader.TrackSyntaxError: Could not load '/Users/rick.boyd/dev/rally-tracks/http_logs/track.json'. The complete track has been written to '/var/folders/_j/86s8k_jd4md2k6ch_290_0mr0000gn/T/tmp8ptuoibm.json' for diagnosis.


```
</details>

**Comment by DJRickyB** [comment]
Hmm... Even in that case it would be nice to get the console to print `jinja2.exceptions.UndefinedError` as the type in the `Caused by:`, which it currently does not.  I'll keep looking at that

**Comment by pquentin** [comment]
You're right, the full traceback is in the logs, just like you've shown in your rally.log snippet. Sorry about that. (I don't know where the empty lines come from though, it would be nice to remove them.)

**Comment by DJRickyB** [comment]
turns out the loss of the type in the console message was due to something specific with how we handle jinja2 exceptions. this has been fixed with the latest commit

File: esrally/track/loader.py [hunk 1]
```diff
@@ -1005,7 +1005,7 @@ def read(self, track_name, track_spec_file, mapping_dir):
             self.logger.exception("Could not load [%s].", track_spec_file)
             msg = "Could not load '{}'. The complete track has been written to '{}' for diagnosis.".format(track_spec_file, tmp.name)
             # Convert to string early on to avoid serialization errors with Jinja exceptions.
```
Comment (pquentin) [inline esrally/track/loader.py:1007]: Can you please remove the comment that becomes outdated by removing the `str()` call?

File: esrally/utils/modules.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def _modules(self, module_paths, component_name):
         for path in module_paths:
             for filename in os.listdir(path):
                 name, ext = os.path.splitext(filename)
-                if ext.endswith(".py"):
+                if ext.endswith(".py") and name != "__init__":
```
Comment (pquentin) [inline esrally/utils/modules.py:55]: Is this related?

File: esrally/actor.py [hunk 3]
```diff
@@ -85,16 +85,27 @@ def receiveMsg_DefuseBomb(self, msg, sender):
...
             # don't forward the exception as is because the main process might not have this class available on the load path
             # and will fail then while deserializing the cause.
-            self.send(sender, BenchmarkFailure("{} ({})".format(msg, str(e))))
+            parsed_exception = _format_exception(e)
+            nesting = 0
+            while hasattr(e, "cause") and e.cause:
+                nesting += 1
+                e = e.cause
+                parsed_exception += f" (Caused by: {_format_exception(e)}"
+            parsed_exception += nesting * ")"
```
Comment (pquentin) [inline esrally/actor.py:107]: What are your thoughts on using the traceback module?

```suggestion
            parsed_exception = traceback.format_exc()
```

It's shorter and simpler. It does print a traceback to stdout, but that's to be expected when you run a Python program from the command line? (Not asking you to change this, just suggesting the idea. I know you didn't want to print everything to stdout initially.)

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks, this type of change is awesome for casual users like me! It works really well. I've left a few questions

File: esrally/track/loader.py (hunk 1 above)
Comment (DJRickyB) [reply esrally/track/loader.py:1007]: yep

File: esrally/utils/modules.py (hunk 2 above)
Comment (DJRickyB) [reply esrally/utils/modules.py:55]: nope

File: esrally/actor.py (hunk 3 above)
Comment (DJRickyB) [reply esrally/actor.py:107]: results in 
```
 % esrally race --track-path ../rally-tracks/http_logs --test-mode

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [c1f21a58-d307-4493-8c37-487d69b6c183]
[ERROR] Cannot race. Traceback (most recent call last):
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 982, in read
    rendered = render_template_from_file(track_spec_file, self.track_params, complete_track_params=self.complete_track_params)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 781, in render_template_from_file
    return render_template(
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 750, in render_template
    return template.render()
  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 48, in top-level template code
jinja2.exceptions.UndefinedError: 'runtime' is undefined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/rick.boyd/dev/rally/esrally/actor.py", line 96, in guard
    return f(self, msg, sender)
  File "/Users/rick.boyd/dev/rally/esrally/racecontrol.py", line 110, in receiveMsg_Setup
    self.coordinator.setup(sources=msg.sources)
  File "/Users/rick.boyd/dev/rally/esrally/racecontrol.py", line 194, in setup
    self.current_track = track.load_track(self.cfg)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 192, in load_track
    return _load_single_track(cfg, repo, repo.track_name)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 199, in _load_single_track
    current_track = reader.read(track_name, track_repository.track_file(track_name), track_dir)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 1007, in read
    raise TrackSyntaxError(msg, e)
esrally.track.loader.TrackSyntaxError: Could not load '/Users/rick.boyd/dev/rally-tracks/http_logs/track.json'. The complete track has been written to '/var/folders/_j/86s8k_jd4md2k6ch_290_0mr0000gn/T/tmpyjca3iqo.json' for diagnosis.


Getting further help:
*********************
* Check the log files in /Users/rick.boyd/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/latest/.
* Ask a question on the forum at https://discuss.elastic.co/tags/c/elastic-stack/elasticsearch/rally.
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /Users/rick.boyd/.rally/logs.

--------------------------------
[INFO] FAILURE (took 30 seconds)
--------------------------------
```
What do you (and @danielmitterdorfer) think?

Comment (DJRickyB) [reply esrally/actor.py:107]: yeah... i'm good with it

**Comment by DJRickyB** [comment]
Thanks @pquentin, I like how much simpler this change is now

File: esrally/actor.py [hunk 4]
```diff
@@ -86,15 +87,16 @@ def receiveMsg_DefuseBomb(self, msg, sender):
...
-            msg = "Error in {}".format(actor_name)
+        except BaseException:
             # log here as the full trace might get lost.
-            logging.getLogger(__name__).exception(msg)
+            logging.getLogger(__name__).exception("Error in %s", actor_name)
             # don't forward the exception as is because the main process might not have this class available on the load path
             # and will fail then while deserializing the cause.
-            self.send(sender, BenchmarkFailure("{} ({})".format(msg, str(e))))
+            parsed_exception = traceback.format_exc()
+            self.send(sender, BenchmarkFailure(parsed_exception))
```
Comment (pquentin) [inline esrally/actor.py:99]: Can we stop using the word `parsed_exception`? It's now a traceback. (I guess we don't need a variable at all.)

**Review by pquentin — APPROVED** [review]
Thanks! Merge when ready. (Unless you want Daniel to take a look.)

I wonder if we should start printing a traceback for the general case too?

File: esrally/actor.py (hunk 3 above)
Comment (pquentin) [reply esrally/actor.py]: (PEP 654 was indeed accepted for Python 3.11: https://discuss.python.org/t/accepting-pep-654-exception-groups-and-except/10813)

**Review by danielmitterdorfer — APPROVED** [review]
Thanks! LGTM
